src/materialized/mux.rs

```diff
@@ -7,132 +7,155 @@
-        // out protocols with small handshakes, like pgwire. Read too few bytes
-        // and you won't be able to tell what protocol you have. For now, eight
-        // bytes is the magic number, but this may need to change if we learn to
-        // speak new protocols.
-        let mut ss = SniffingStream::new(conn);
-        let mut buf = [0; 8];
-        let nread = match netio::read_exact_or_eof(&mut ss, &mut buf).await {
-            Ok(nread) => nread,
-            Err(err) => {
-                error!("error handling request: {}", err);
-                return;
-            }
-        };
-        let buf = &buf[..nread];
-
-        for handler in &self.handlers {
-            if handler.match_handshake(buf) {
-                if let Err(e) = handler.handle_connection(ss.into_sniffed()).await {
-                    error!("error handling connection: {}", e);
+    /// Serves incoming TCP traffic from `listener`.
+    pub async fn serve(self, mut incoming: Incoming<'_>) {
+        let handlers = Arc::new(self.handlers);
+        while let Some(conn) = incoming.next().await {
+            let conn = match conn {
+                Ok(conn) => conn,
+                Err(err) => {
+                    error!("error accepting connection: {}", err);
+                    continue;
                 }
-                return;
-            }
+            };
+            // Set TCP_NODELAY to disable tinygram prevention (Nagle's
+            // algorithm), which forces a 40ms delay between each query
+            // on linux. According to John Nagle [0], the true problem
+            // is delayed acks, but disabling those is a receive-side
+            // operation (TCP_QUICKACK), and we can't always control the
+            // client. PostgreSQL sets TCP_NODELAY on both sides of its
+            // sockets, so it seems sane to just do the same.
+            //
+            // If set_nodelay fails, it's a programming error, so panic.
+            //
+            // [0]: https://news.ycombinator.com/item?id=10608356
+            conn.set_nodelay(true).expect("set_nodelay failed");
+            tokio::spawn(handle_connection(handlers.clone(), conn));
+        }
+    }
+}
+
+async fn handle_connection(handlers: Arc<Handlers>, conn: TcpStream) {
+    // Sniff out what protocol we've received. Choosing how many bytes to
+    // sniff is a delicate business. Read too many bytes and you'll stall
+    // out protocols with small handshakes, like pgwire. Read too few bytes
+    // and you won't be able to tell what protocol you have. For now, eight
+    // bytes is the magic number, but this may need to change if we learn to
+    // speak new protocols.
+    let mut ss = SniffingStream::new(conn);
+    let mut buf = [0; 8];
+    let nread = match netio::read_exact_or_eof(&mut ss, &mut buf).await {
+        Ok(nread) => nread,
+        Err(err) => {
+            error!("error handling request: {}", err);
```

Comment:
micronit, tioli: this will show up in logs with a bit of a stutter `ERROR: error handling request: {}`, maybe rephrasing to `unable to handle new connection: {}`?

Reply:
Good point, thanks! Fixed.